nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,561 @@
+
+    def get_nx_graph(self):
+        return self._nx_graph
+
+def prepare_search_graph(original_graph: PTNNCFGraph) -> SearchGraph:
+    nx_merged_graph = get_orig_graph_with_orig_pattern(original_graph.get_nx_graph_copy())
+    sgraph = SearchGraph(nx_merged_graph)
+    return sgraph
+
+def get_orig_graph_with_orig_pattern(orig_graph):
+    # pylint: disable=protected-access
+    pattern_fusing_graph = PT_HW_FUSED_PATTERNS.get_full_pattern_graph()
+    matches = find_subgraphs_matching_pattern(orig_graph, pattern_fusing_graph)
+    merged_graph = deepcopy(orig_graph)
+    nx.set_node_attributes(merged_graph, False, SearchGraph.IS_DUMMY_NODE_ATTR)
+    nx.set_node_attributes(merged_graph, False, SearchGraph.IS_MERGED_NODE_ATTR)
+    nx.set_node_attributes(merged_graph, None, SearchGraph.ACTIVATION_INPUT_SHAPE_ATTR)
+    nx.set_node_attributes(merged_graph, None, SearchGraph.ACTIVATION_OUTPUT_SHAPE_ATTR)
+    for match in matches:
+        if len(match) == 1:
+            continue
+        input_node_key = match[0]
+        output_node_key = match[-1]
+        in_edges = list(merged_graph.in_edges(input_node_key))
+        out_edges = list(merged_graph.out_edges(output_node_key))
+
+        in_edge_copies_dict = {}
+        for in_edge_key in in_edges:
+            in_edge_copies_dict[in_edge_key] = deepcopy(merged_graph.edges[in_edge_key])
+        out_edge_copies_dict = {}
+        for out_edge_key in out_edges:
+            out_edge_copies_dict[out_edge_key] = deepcopy(merged_graph.edges[out_edge_key])
+
+        merged_node_key = ""
+        merged_nodes = []
+        type_list = []
+        for node_key in match:
+            attrs = orig_graph.nodes[node_key]
+            merged_node_key +=  str(attrs['id']) + ' ' + attrs[SearchGraph.TYPE_NODE_ATTR] + '  '
+            # pylint: disable=protected-access
+            merged_nodes.append(orig_graph.nodes[node_key])
+            merged_graph.remove_node(node_key)
+            type_list.append(attrs[SearchGraph.TYPE_NODE_ATTR])
+        merged_node_attrs = {
+            SearchGraph.KEY_NODE_ATTR: merged_node_key,
+            SearchGraph.IS_MERGED_NODE_ATTR: True,
+            SearchGraph.TYPE_NODE_ATTR: type_list,
+            SearchGraph.MERGED_NODES_NODE_ATTR: merged_nodes,
+        }
+        merged_graph.add_node(merged_node_key, **merged_node_attrs)
+        for in_edge_key, in_edge_attrs in in_edge_copies_dict.items():
+            merged_graph.add_edge(in_edge_key[0], merged_node_key, **in_edge_attrs)
+        for out_edge_key, out_edge_attrs in out_edge_copies_dict.items():
+            merged_graph.add_edge(merged_node_key, out_edge_key[1], **out_edge_attrs)
+
+    return merged_graph
+
+def add_node_to_aux_struct(node_key, shape, shape_map):
+    str_shape = str(shape)
+    if str_shape in shape_map:
+        shape_map[str_shape].add(node_key)
+    else:
+        shape_map[str_shape] = set([node_key])
+
+def rule_after_removing_block_graph_has_no_hanging_edges(graph: SearchGraph,
```

Comment:
The function name has to reflect what it _does_, i.e. should ideally contain a verb. `rule` here and in functions below is hardly a verb.

Reply:
Done.